ThirdPartyResource example: added watcher example, code cleanup

nilebox opened the pull request:
**NOTE**: This is a cleaned and updated version of PR https://github.com/kubernetes/kubernetes/pull/43027

**What this PR does / why we need it**:
An example of using go-client for watching on ThirdPartyResource events (create/update/delete).

Commit:
ThirdPartyResource client-go example: added TPR controller example, code cleanup and integration test

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45463)
<!-- Reviewable:end -->

nilebox commented:
@sttts @caesarxuchao Please see this updated PR.
I have removed the unneccessary and contradictory code which has blocked the previous PR from being merged.
As there is a duplicate https://github.com/kubernetes/kubernetes/pull/45263, it proves the importance of such an example.

MikeSpreitzer commented on staging/src/k8s.io/client-go/examples/third-party-resources/main.go:
```diff
@@ -52,24 +50,29 @@ func main() {
+
+			// See the issue https://github.com/kubernetes/features/issues/95
+			// ("Make new TPRs available immediately after the create request succeeds")
+			fmt.Print("Sleeping to make sure the TPR is processed and available")
```
How about upgrading the comment, if not the printed message, to point to the existing issue (https://github.com/kubernetes/features/issues/95) that notes the need for a delay (among many other things, sigh)?

nilebox replied:
@MikeSpreitzer It does have a comment above pointing to this issue already :)
Not sure if adding the issue link to the printed message helps...

MikeSpreitzer reviewed (commented):
If we are trying to give a good example, that would use glog instead of fmt, right?
But that does not need to hold up this PR, it could be done in a later PR.

sttts commented:
@k8s-bot ok to test

sttts commented:
@nilebox will take a look. In general, try to avoid creating new PRs as this remove the context of previous discussions.

sttts commented:
@nilebox am still not really happy with the example. IMO we have to simplify it quite a lot. Let me contribute an integration test on-top of your code. This is far too complex to be without test.

nilebox commented:
@sttts sure, integration test would be helpful.

sttts commented on staging/src/k8s.io/client-go/examples/third-party-resources/controller.go:
```diff
@@ -0,0 +1,104 @@
+	return controller, nil
+}
+
+func onAdd(obj interface{}) {
```
usually these are methods of the controller.

sttts commented on staging/src/k8s.io/client-go/examples/third-party-resources/main.go:
```diff
@@ -52,66 +53,65 @@ func main() {
+	// In the normal application there is no need for it, because:
+	// 1. It's unlikely to create a watcher and a TPR instance at the same time in the same application.
+	// 2. The application with watcher would most probably keep running instead of exiting right after the watcher startup.
+	time.Sleep(5 * time.Second)
```
Do we pass a zero resourceVersion to the Watch request? We should get all old events that are in the apiserver watch cache.

nilebox replied:
Yes, I think it's fine to get all the old events?
This sleep is only to ensure that we handle the example created below (on the first run, when we don't have any old objects and will create a new one).

sttts commented on staging/src/k8s.io/client-go/examples/third-party-resources/main.go:
```diff
@@ -52,66 +53,65 @@ func main() {
 
 	if err != nil {
-		if errors.IsNotFound(err) {
+		if apierrors.IsNotFound(err) {
```
Just try to post the tpr and check for errors.IsAlreadyExists. No need for the whole Get+Post logic here.

sttts replied:
Compare the tpr creation itself. We do the same there.

sttts commented on test/integration/thirdparty/client-go_test.go:
```diff
@@ -0,0 +1,57 @@
+		t.Fatalf("ThirdPartyResource examples did not show up: %v", err)
+	}
+	t.Logf("TPR %q is active", exampletprv1.ExampleResourcePlural)
+
```
Here we need the controller as well. For that you have to move the ExampleController into a `controller` sub-package.

caesarxuchao commented on test/integration/thirdparty/client-go_test.go:
```diff
@@ -0,0 +1,73 @@
+	if err != nil {
+		t.Fatalf("unexpected error starting a TPR controller: %v", err)
+	}
+
```
nit: remove the empty lines.

caesarxuchao commented:
/release-note-none

caesarxuchao commented:
Thanks @nilebox! @sttts could you apply the lgtm when you approve it? It generally looks good to me. I don't have time to review closely. Thanks.

caesarxuchao commented on staging/src/k8s.io/client-go/examples/third-party-resources/client/client.go:
```diff
@@ -0,0 +1,45 @@
+
+func NewClient(cfg *rest.Config) (*rest.RESTClient, *runtime.Scheme, error) {
+	scheme := runtime.NewScheme()
+	if err := tprv1.AddToScheme(scheme); err != nil {
```
Is List/Watch working for tpr client? I guess they are since the controller needs them but i don't understand why they are working. List/Watch are using the global scheme/codec, see this [line](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/kubernetes/typed/core/v1/pod.go#L156), not the scheme constructed here. I thought the global scheme/codec would complain not recognizing the tpr groupVersionKind.

caesarxuchao replied:
Nevermind, the apimachinery code is very tolerant for unversioned Kind (ListOptions is registered as one) and thus there is no error.

nilebox replied:
@caesarxuchao yeah, looks like there were some changes recently, it was breaking before but works fine now, at least for the code used in the example.

ash2k commented on staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go:
```diff
@@ -0,0 +1,102 @@
+		})
+
+	// store can be used to List and Get
+	for _, obj := range store.List() {
```
`store.List()` will always return an empty slice here because Informer has not been started yet and hadn't done any list/watch operations yet. Maybe add `cache.WaitForCacheSync(ctx.Done(), controller.HasSynced)` after the `controller.Run()` to make sure it has synced and move this `for` after it?

nilebox replied:
idk, maybe just remove this code to keep the PR simple...

ash2k commented on staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go:
```diff
@@ -0,0 +1,102 @@
+	source := cache.NewListWatchFromClient(
+		c.ExampleClient,
+		tprv1.ExampleResourcePlural,
+		apiv1.NamespaceAll,
```
From what I understand from other code, the convention is to use `metav1.Namespace*` constants? `apiv1.Namespace*` constants are only for the package where they reside?

nilebox replied:
Not sure, all the examples using `NewListWatchFromClient()` I've found are invoking `apiv1.NamespaceAll`

ash2k commented on staging/src/k8s.io/client-go/examples/third-party-resources/controller/controller.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+func (c *ExampleController) onDelete(obj interface{}) {
+	example := obj.(*tprv1.Example)
```
It would be awesome for completeness to show that this may fail sometimes - see https://github.com/atlassian/smith/pull/54 for example. I believe that is what all listeners in Kubernetes codebase do.

nilebox replied:
Good point, but I think it should be a separate PR. This one is already XL...

ash2k commented on staging/src/k8s.io/client-go/examples/third-party-resources/main.go:
```diff
@@ -52,83 +54,63 @@ func main() {
+	var result tprv1.Example
+	err = exampleClient.Post().
+		Resource(tprv1.ExampleResourcePlural).
+		Namespace(apiv1.NamespaceDefault).
```
`metav1.NamespaceDefault`?

ash2k commented on test/integration/thirdparty/client-go_test.go:
```diff
@@ -0,0 +1,73 @@
+
+	config := &rest.Config{Host: s.URL, ContentConfig: rest.ContentConfig{NegotiatedSerializer: api.Codecs}}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
```
nit: I think at least some other packages use `assert.NoError(err)`/`require.NoError(err)` to reduce boilerplate.

nilebox commented:
@sttts I have added a bunch of code for testing that controller actually processes the Example instance by changing its status. Also, it helped me replacing sleeps with polling for expected condition. Added this to an integration test as well, please review.

sttts commented on staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go:
```diff
@@ -23,25 +23,49 @@ import (
+	Metadata metav1.ObjectMeta `json:"metadata"`
 
 	Spec ExampleSpec `json:"spec"`
+
```
nit: no empty line

sttts commented on staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go:
```diff
@@ -23,25 +23,49 @@ import (
-	Metadata        metav1.ObjectMeta `json:"metadata"`
+	// WARNING: Don't call the field below ObjectMeta to avoid the issue with ugorji
+	// https://github.com/kubernetes/client-go/issues/8#issuecomment-285333502
+	Metadata metav1.ObjectMeta `json:"metadata"`
```
@nilebox in my last PR I had removed this Metadata and everything kept working. I doubt we need this workaround.

nilebox replied:
@sttts completely removed? But having metadata (a field of ObjectMeta types) is useful, why would we remove it? Also, Metadata fields are present in current master (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/examples/third-party-resources/types.go#L33).

nilebox replied:
@sttts I think you meant keep it named ObjectMeta/ListMeta. Just checked, and yes, it seems to work now... Also it means there is no need of implementing `GetObjectKind` and other methods. Looks like some recent change has fixed this issue.
Will get rid of it then.

nilebox replied:
@sttts looks like we don't need other ugorji workarounds below either. Awesome.

sttts replied:
I meant to embed the ObjectMeta directly, without `Metadata`.

sttts commented on staging/src/k8s.io/client-go/examples/third-party-resources/apis/tpr/v1/types.go:
```diff
@@ -23,25 +23,49 @@ import (
-	Metadata        metav1.ListMeta `json:"metadata"`
+	// WARNING: Don't call the field below ListMeta to avoid the issue with ugorji
+	// https://github.com/kubernetes/client-go/issues/8#issuecomment-285333502
+	Metadata metav1.ListMeta `json:"metadata"`
```
same here

nilebox commented:
@sttts all the checks succeeded except "Jenkins verification". Not sure why it fails.

nilebox commented:
@sttts all checks succeded, need lgtm and squashing commits

sttts commented:
lgtm. Please squash and let me know when I should label.

nilebox commented:
@sttts done.

sttts commented:
Self-approving due to https://github.com/kubernetes/kubernetes/pull/45463#issuecomment-299988184

nilebox commented:
@k8s-bot unit test this
